dap/adapter.go

```diff
@@ -0,0 +1,455 @@
+		Event: dap.Event{Event: "thread"},
+		Body: dap.ThreadEventBody{
+			Reason:   "exited",
+			ThreadId: t.id,
+		},
+	}
+}
+
+type evaluateRequest struct {
+	name  string
+	c     gateway.Client
+	ref   gateway.Reference
+	meta  map[string][]byte
+	errCh chan<- error
+}
+
+func (d *Adapter) EvaluateResult(ctx context.Context, name string, c gateway.Client, res *gateway.Result) error {
+	eg, _ := errgroup.WithContext(ctx)
+	if res.Ref != nil {
+		eg.Go(func() error {
+			return d.evaluateRef(ctx, name, c, res.Ref, res.Metadata)
+		})
+	}
+
+	for k, ref := range res.Refs {
+		refName := fmt.Sprintf("%s (%s)", name, k)
+		eg.Go(func() error {
+			return d.evaluateRef(ctx, refName, c, ref, res.Metadata)
+		})
+	}
+	return eg.Wait()
+}
+
+func (d *Adapter) evaluateRef(ctx context.Context, name string, c gateway.Client, ref gateway.Reference, meta map[string][]byte) error {
+	errCh := make(chan error, 1)
+
+	// Send a solve request to the launch routine
+	// which will perform the solve in the context of the server.
+	ereq := &evaluateRequest{
+		name:  name,
+		c:     c,
+		ref:   ref,
+		meta:  meta,
+		errCh: errCh,
+	}
+	select {
+	case d.evaluateReqCh <- ereq:
+	case <-ctx.Done():
+		return context.Cause(ctx)
+	}
+
+	// Wait for the response.
+	select {
+	case err := <-errCh:
+		return err
+	case <-ctx.Done():
+		return context.Cause(ctx)
+	}
+}
+
+func (d *Adapter) Threads(c Context, req *dap.ThreadsRequest, resp *dap.ThreadsResponse) error {
+	d.threadsMu.RLock()
+	defer d.threadsMu.RUnlock()
+
+	for _, t := range d.threads {
+		resp.Body.Threads = append(resp.Body.Threads, dap.Thread{
+			Id:   t.id,
+			Name: t.name,
+		})
+	}
+	return nil
+}
+
+func (d *Adapter) StackTrace(c Context, req *dap.StackTraceRequest, resp *dap.StackTraceResponse) error {
+	t := d.getThread(req.Arguments.ThreadId)
+	if t == nil {
+		return errors.Errorf("no such thread: %d", req.Arguments.ThreadId)
+	}
+
+	resp.Body.StackFrames = t.StackFrames()
+	return nil
+}
+
+func (d *Adapter) evaluate(ctx context.Context, name string, c gateway.Client, res *gateway.Result) error {
+	errCh := make(chan error, 1)
+
+	started := d.srv.Go(func(ctx Context) {
+		defer close(errCh)
+		errCh <- d.EvaluateResult(ctx, name, c, res)
+	})
+	if !started {
+		return context.Canceled
+	}
+
+	select {
+	case err := <-errCh:
+		return err
+	case <-ctx.Done():
+		return context.Cause(ctx)
+	}
+}
+
+func (d *Adapter) Handler() build.Handler {
+	return build.Handler{
+		Evaluate: d.evaluate,
+	}
+}
+
+func (d *Adapter) dapHandler() Handler {
+	return Handler{
+		Initialize:        d.Initialize,
+		Launch:            d.Launch,
+		Continue:          d.Continue,
+		SetBreakpoints:    d.SetBreakpoints,
+		ConfigurationDone: d.ConfigurationDone,
+		Threads:           d.Threads,
+		StackTrace:        d.StackTrace,
+		Evaluate:          d.Evaluate,
+	}
+}
+
+type thread struct {
+	d    *Adapter
+	id   int
+	name string
+
+	paused chan struct{}
+	rCtx   *build.ResultHandle
```

Comment:
follow-up: this `ResultHandle` seems incorrectly named. I'm not sure if we need it at all but if we do then it should be called something else and should not be defined in the `build` pkg, as it is basically the transformation of the `Evaluate()` parameters into struct form and therefore duplicate in the context of build pkg.

`Ctx` variable naming also makes it confusing as it looks like some form of `context.Context`.

Reply:
Agreed. It's mostly been kept around because the code exists and it was useful to reuse it. It can be renamed and probably should be renamed.